### PR TITLE
[IMP] Move config generation to entrypoint

### PR DESCRIPTION
**Pull request by lasley**
I'm pretty sure this is the only change required for #37, although I'll test later today. 

I still need to update the docs and all that, but I figured I'd put my code where my mouth is - particularly seeing as I need it this way anyways 😆 

I left the possibility of generating the config at build so that we don't kill your images, although I recommend we leave the default to build on entrypoint instead. This means that you have to edit all your images though, which kind of sucks, but the other way is insecure by default which I think is bad.

### CODE REVIEW

**Comment by lasley** [comment]
Seems this works too - hooray!

File: build.d/60-config-generate [hunk 1]
```diff
@@ -1,9 +1,8 @@
...
-set -e
-src="/opt/odoo/common/conf.d/* /opt/odoo/custom/conf.d/*"
-log INFO Merging $(ls $src | wc -l) configuration files in $OPENERP_SERVER
-conf=$(cat $src | envsubst)
-log DEBUG "Resulting configuration:
-$conf"
-echo "$conf" > $OPENERP_SERVER
+
+if [ "${CONFIG_BUILD}" != true ]; then
+    log WARNING Skipping Odoo config generation - will perform during entrypoint.
```
Comment (Yajo) [inline build.d/60-config-generate]: If this is the new normal behavior, then the warning should go in the opposite case.

File: entrypoint.d/50-config-generate [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+if [ "${CONFIG_BUILD}" = true ]; then
```
Comment (Yajo) [inline entrypoint.d/50-config-generate]: This env will never be available at run time (notice it is an `ARG`, not an `ENV` in the Dockerfile).
Instead, you can just check for existence of the file indicated in `$OPENERP_SERVER`.

**Comment by lasley** [comment]
This looks to be working in my tests

**Comment by Yajo** [comment]
But one question... when you reboot Odoo, do you want a config file to be regenerated? I think that would make most sense, and in such case, we should try to keep it simple and remove the thing from build and move it to entrypoint, no less, no more.

**Comment by lasley** [comment]
I agree we want it generated every time. Honestly the only reason I left it in build was for backwards compatibility, but if you're good with not supporting it so am I. All my scaffolds use environment vars, so it doesn't affect me really.

File: Dockerfile [hunk 3]
```diff
@@ -16,6 +16,7 @@ ONBUILD ARG DEPTH_DEFAULT=1
 ONBUILD ARG DEPTH_MERGE=100
 ONBUILD ARG CLEAN=true
 ONBUILD ARG COMPILE=true
+ONBUILD ARG CONFIG_BUILD=false
```
Comment (Yajo) [inline Dockerfile]: What's this for then?

File: entrypoint.d/50-config-generate [hunk 4]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+config-generate
```
Comment (Yajo) [inline entrypoint.d/50-config-generate]: Why not just put the file here? I see no more use cases...

**Comment by Yajo** [comment]
I like it now 😊 

Can you fix conflicts and squash please?

**Comment by lasley** [comment]
@Yajo done

**Comment by Yajo** [comment]
This one broke many things, I reverted it at #48.

Sorry to say, but merging nothing more until #43 is fixed. 😞

**Comment by lasley** [comment]
@Yajo - what did it break? I had already warned that essentially all of your scaffolding and existing images would be broken with this in #37 & it seemed you guys were good with this.

As it stands now, this repo cannot scale past 1:1 image:customer. This is an absolute blocker for use of this repo, as well as a huge security issue storing the passwords for the database and instance admin passwords in the images.

**Comment by Yajo** [comment]
Yeah, it's not your fault, it's mine for merging without proper testing, but I had to roll it all back in a rush. 😅 

We want this feature, but cannot merge without proper backwards compatibility.

The main problem was that older deployments had build args where new ones would need env vars. Build args do not get converted into env vars unless explicitly made in the Dockerfile. That's most likely the backwards compatibility layer that we will need to maintain for awhile. Something like:

```Dockerfile
ONBUILD ARG WITHOUT_DEMO=all
ONBUILD ENV WITHOUT_DEMO="$WITHOUT_DEMO"
```

... the same as that for every variable used under `common/conf.d`. This way if you use env vars, default ones get overriden at runtime through your docker-compose file, but if you have an old deployment using build args, those get converted into env vars at build time that can be used at runtime to get the config file generated.

As you can see, not much hassle, but I had no time to fix & test properly and I had to roll it back or we were going to get broken builds everywhere.

Let's just reopen this in a new PR and continue from there if you want. 😉 

> As it stands now, this repo cannot scale past 1:1 image:instance.

I don't understand what you mean...

> huge security issue storing the passwords for the database and instance admin passwords in the images.

Unless you use secrets, it's not much different from storing them in runtime, but using secrets involves fixing #20, which is in the roadmap, but is not our current utmost priority.

**Comment by lasley** [comment]
> As it stands now, this repo cannot scale past 1:1 image:instance.
>> I don't understand what you mean...

The Docker images that are built contain an admin and a Postgres password. Unless you plan on sharing passwords between your customer instances, this is a bottleneck to vertical scaling. IMO database name & user is too.

> Unless you use secrets, it's not much different from storing them in runtime, but using secrets involves fixing #20, which is in the roadmap, but is not our current utmost priority.

It's much different. My container infrastructure is all in HIPAA compliant, audited infrastructure. Build images are stored either in Docker or AWS Container Registry. Neither of these services are secure, and thus my passwords have been exposed across a known insecure plane.

**Comment by Yajo** [comment]
OK now I get your points, thanks for the explanation. So now the goal is to fix #43 and then add this feature again 😊

**Pull request merged**